### ISSUE TITLE
porntube.com [NSFW][missing filters]

### ISSUE THREAD

**Issue by WowDude** [uBlockOrigin/uAssets#1821]
As for https://github.com/uBlockOrigin/uAssets/issues/1774#issuecomment-376124219 and later comments.

> @gorhill : @WowDude OP issue is solved, leave it alone please. If you think the script should be blocked make the case elsewhere

OK, as you wish.

> @gorhill : Also, do not embed pornographic images here, use text-only URLs and mark them clearly as NSFW.

- 1 just didn't realise
- 2 thought that putting NSFW in the title + putting the porn image into colapsed spoiler is enough
- 3 btw. it seems that there is not much difference between viewing the porn image by uncollapsing a spoiler vs by visiting a code-tagged link. In the both cases the image is hidden, and requires a manual action to unhide it, and still the screenshoot in code-tagged link is hosted on github server
- 4 don't see what you have said about NSFW neither in README.md nor in CONTRIBUTING.md, perhaps I'm missing something and perhaps you provide me link where you stated about such rules about NFSW.
- 5 anyway I'll adapt to your recommendations

### URL(s) where the issue occurs

NSFW `https://www.porntube.com/videos/gia-paige-interracial-muff-banging_7414233`

### Describe the issue

I already described the issue [here](https://github.com/uBlockOrigin/uAssets/issues/1774#issuecomment-376124219), so I copy&paste the most essential part:

I see useless `mojon.js` script loaded which unnecessary wastes resources (bandwidth, CPU, memory/ram, electrical energy etc.) especially that it gets loaded several (6) times.

Unnecessary wasted resources:
- a) bandwidth, memory/ram wasted (file was downloaded and stored): `mojon.js` file size: 24,6 KB 
(the file was requested 6 times, not sure if that counts as 6 x 24,6 KB, as it could be cached)

- b) page load time with mojon.js blocked by `||porntube.com/mojon.js` filter drastically decerased (at least on Opera and probably on Chrome/Chrome based browsers too, I just don't use it), benchmarked using [your benchmark tool](http://www.raymondhill.net/ublock/pageloadspeed.html):

```
Opera without filter:   Average (valid): 1716.55 ms (11/14 iterations)  Last reload: 2136.00 ms
Opera with filter:      Average (valid):  708.00 ms (11/22 iterations)  Last reload: 684.00 ms

Aboout: almost 2.5x difference, 1s gain

Firefox without filter: Average (valid): 1641.82 ms (11/14 iterations)  Last reload: 1634.00 ms
Firefox with filter:    Average (valid): 1498.91 ms (11/28 iterations)  Last reload: 1582.00 ms  

About: almost 1.1x difference, 0,15s gain

(not sure what happens on Firefox, high randomness, I'm a bit surprised, sometimes 
it load in like 0,8s and sometimes 1,6s, perhaps some other factor causing such fluctuations)
```

- c) some ammount of CPU cycles (file must be processed), so possible some ammount of electrical energy/mobile battery drain too

> @gorhill : do not use unrelated quotes which are not specifically related to what you think is an issue.

OK, but the part you are reffering to, was general talking about not wasting resources, it was just a complement to the:

> @gorhill : actual specific evidence

 I provided (not so detailed ok I agree) at the beginning of my comment.

### Screenshot(s)

![](https://user-images.githubusercontent.com/37474031/37901335-2cb328f0-30f1-11e8-9bd5-be30cc5bb293.png)

### Versions

- Browser/version: (O 52.0.2871.30  64bit / F 59.02 64bit
- uBlock Origin version: uBO 1.15.18

### Settings

- Default.

### Notes

As seen, the noticeable gain of resource saving has been proven. And it's rather about generally not wasting resources and blocking unnessecary trash files (on all websites) as well. It doesn't really seem to matter how much resources were wasted, no need to generally waste resources, does it requires to do scientific benchmarks and calcualtions for every single website to deal with a simple fact that blocking such trash files helps? Perhaps benchmark like 20 websites, and if the most of the results will confirm noticeable gain, then just begin to block such trash files in uBlock filters list by filter creators as default. Even if sometimes on some particular websites, blocking such unnessecary objects won't save resources, it's clear that on most webpages it will does. Sometimes it can be a single trash file wasting resources, and sometimes several trash files.

**Comment by gorhill** [comment]
>     Opera without filter:   Average (valid): 1716.55 ms (11/14 iterations)  Last reload: 2136.00 ms
>     Opera with filter:      Average (valid):  708.00 ms (11/22 iterations)  Last reload: 684.00 ms

I get no difference with and without the filter with Chromium, ~900 ms. The _Performance_ tool in Chromium confirm this -- no difference.

If you want that script blocked, you need to find out what the script does, and given its purpose if it's candidate for one of the existing lists, and if blocking it will break things.

**Comment by WowDude** [comment]
> @gorhill : I get no difference with and without the filter with Chromium, ~900 ms. The Performance tool in Chromium confirm this -- no difference.

Nor I.  ~1300ms. 
 
I also have repeated the test on Opera and got again similiar results like before:
```
Average (valid): 1503.08 ms (11/14 iterations)  Last reload: 1407.50 ms
Average (valid):  551.84 ms (11/14 iterations)  Last reload:  576.00 ms
```


Interesting: 
Opera: big gain, 
Firefox: small gain,
 Chrome: no gain: 
2 browsers gain, 
1 browser no gain, 

Well, perhaps some more people should get involved and do some benchmarks, and also some more webpages should be benchmarked, however as seen, there is gain in 2 of 3 browsers for me, and still these unblocked junk files always take memory/ram/bandwidth/cpu/electricity/battery regardless of page speed load time, there is almost always some kind of gain by blocking them.




> @gorhill : You need to find out what the script does

As far as I see it contains nothing more than ExoLoader crap junks.

> @gorhill : and if blocking it will break things.

As far as I see it doesn't break anything, like video playing or browsing the webpage.

**Comment by gorhill** [comment]
> it contains nothing more than ExoLoader crap junks

Well then for me that would be the _proper_ rationale to block -- it matches the purpose of the `filters.txt` list.

**Comment by mapx-** [comment]
There is another problem: those scripts are easy blockable, but sometimes the webmasters are renaming them just to break some functionality and force the user to disable the adblocker (whitelisting the site).

**Comment by gorhill** [comment]
Looks like that script is to work around blockers preventing access to `window.ExoLoader`. What we could do is to defuse the purpose of the script rather than block it:

    porntube.com##script:inject(addEventListener-defuser.js, getexoloader)

I suppose `getexoloader` is less likely to change than the script file name, which is not important for the functionality.

**Comment by mapx-** [comment]
However, the existent filter is already working
`porntube.com##script:inject(abort-on-property-read.js, ExoLoader.serve)`

**Comment by mapx-** [comment]
@WowDude  in opera have you disabled the builtin adblocker ?

**Comment by WowDude** [comment]
@mapx- sure, I have it disabled

Btw. I see such filters blocked in `Adguard Base Filters`
```
||fux.com/mojon.js
||4tube.com/mojon.js
```

**Comment by gorhill** [comment]
The thing is the page loads many `iframe` and fetch an out of scope `ExoLoader` through an event listener.  The existing filter takes care of `window.ExoLoader`, not the out of scope ones. Another approach would be to also outright block the iframes:

    ||porntube.com/api/iframe/$subdocument

These are the ones loading `mojon.js`.

**Comment by WowDude** [comment]
@gorhill I have applied your filter and it generally works - it blocks 5 of 6 requests to `mojon.js` but still a single `mojon.js` is loading (i see it in the logger).

**Comment by WowDude** [comment]
@mapx- the same, blocks 5 of 6, but still a single `mojon.js` loading

**Comment by mapx-** [comment]
I know, I'm asking if you still gain something in opera (blocking mojon crap) if you use my filters ?

**Comment by WowDude** [comment]
@mapx- : Yes I do gain: `Average (valid): 581.52 ms (11/15 iterations)  Last reload: 571.40 ms`

**Comment by mapx-** [comment]
So - if I understood correctly - now in opera is small gain - like FF ? in this  case is preferable a "generic" $csp filter.

**Comment by WowDude** [comment]
Also I do the same gain with @gorhill ' s filter: `||porntube.com/api/iframe/$subdocument`
`Average (valid): 581.99 ms (11/18 iterations)  Last reload: 576.30 ms`

> @mapx- : So - if I understood correctly - now in opera is small gain - like FF ? in this case is preferable a "generic" $csp filter.

Nope, what I pasted is an average page load time, not a gain, the gain is the big gain like 1 s because without filters it takes about 1.5s to load the webpage, with filters 0.5s, so it gives a 1s gain.

**Comment by mapx-** [comment]
I was saying you have small gain in opera blocking mojon script over using my filters (or gorhill filters), right ? In this case I'll add the $csp filter.

**Comment by WowDude** [comment]
I understand what you mean, but nope there is almost no difference (I guess that what you see is is just an error margin):
1
`||porntube.com/mojon.js`

2 
`||porntube.com/api/iframe/$subdocument`

3 
`||porntube.com^$csp=child-src 'none'`
`porntube.com##iframe`

```
1 Average (valid): 581.99 ms (11/18 iterations)  Last reload: 576.30 ms
2 Average (valid): 581.52 ms (11/15 iterations)  Last reload: 571.40 ms
3 Average (valid): 551.84 ms (11/14 iterations)  Last reload: 576.00 ms
```

**Comment by gorhill** [comment]
> In this case I'll add the $csp filter.

Well there could be legitimate iframe in there sometimes (maybe at log in time). I think safest is to target the specific `api/iframe` iframes and see how it goes from there.

**Comment by okiehsch** [comment]
I added
`porntube.com##script:inject(set-constant.js, adsEnabled, true)`
which prevents the `porntube.com/api/iframe/` requests.

**Comment by WowDude** [comment]
And one thing as for https://github.com/uBlockOrigin/uAssets/issues/1774#issuecomment-376389509 , it seems you are so rude to give me orders arrogant tone like to a dog, so let me give you several too:
- don't create a drama where nothing bad happened 
- don't lock/break conversation where nothing bad happened
- don't take people the opportunity to talk calmly
- don't make me look like the one who creates a drama, because I just talked to @mapx- , and it was you who panicked and dramatized and was defending him for no reason, as if I was attacking him
- don't talk to me such an arrogant tone like to a dog nor break / lock the conversation this way, you stupid rude son of a bitch.

So what you gonna do now, probably again say some crap on me and lock the conversation, now, fetch dog.
